On Windows .shouldStillHaveLastSnapshotIfFailToMoveTemporaryFile fails every time

On Windows the test FsTemporarySnapshotWriterTest.shouldStillHaveLastSnapshotIfFailToMoveTemporaryFile fails on every run. (Tested with intellij and maven).

```java
java.lang.AssertionError: 
Expecting file:
  <C:\Users\zell\AppData\Local\Temp\junit2135811262865991534\test1239489485.tmp>
not to exist

	at io.zeebe.logstreams.fs.snapshot.FsTemporarySnapshotWriterTest.shouldStillHaveLastSnapshotIfFailToMoveTemporaryFile(FsTemporarySnapshotWriterTest.java:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

@ThorbenLindhauer can you verify that?

BTW: The file exist

--- a/broker-core/src/main/java/io/zeebe/broker/clustering/base/raft/RaftPersistentConfiguration.java
+++ b/broker-core/src/main/java/io/zeebe/broker/clustering/base/raft/RaftPersistentConfiguration.java
@@ -85,8 +85,8 @@ public class RaftPersistentConfiguration implements RaftPersistentStorage {
   }
 
   public void delete() {
-    file.delete();
-    tmpFile.delete();
+    FileUtil.deleteFile(file);
+    FileUtil.deleteFile(tmpFile);
   }
 
   @Override

--- a/broker-core/src/test/java/io/zeebe/broker/exporter/jar/ExporterJarRepositoryTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/exporter/jar/ExporterJarRepositoryTest.java
@@ -23,6 +23,7 @@ import static org.junit.Assume.assumeTrue;
 
 import io.zeebe.broker.exporter.util.JarCreatorRule;
 import io.zeebe.broker.exporter.util.TestJarExporter;
+import io.zeebe.util.FileUtil;
 import java.io.File;
 import java.io.IOException;
 import org.junit.Rule;
@@ -69,7 +70,7 @@ public class ExporterJarRepositoryTest {
     final File dummy = temporaryFolder.newFile("missing.jar");
 
     // when
-    dummy.delete();
+    FileUtil.deleteFile(dummy);
 
     // then
     assertThatThrownBy(() -> jarRepository.load(dummy.getAbsolutePath()))

--- a/broker-core/src/test/java/io/zeebe/broker/exporter/jar/ExporterJarRepositoryTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/exporter/jar/ExporterJarRepositoryTest.java
@@ -52,11 +52,10 @@ public class ExporterJarRepositoryTest {
   @Test
   public void shouldThrowExceptionOnLoadIfNotReadable() throws Exception {
     // given
-    assumeTrue(!System.getProperty("os.name").startsWith("Windows"));
     final File dummy = temporaryFolder.newFile("unreadable.jar");
 
-    // when
-    dummy.setReadable(false);
+    // when (ignoring test if file cannot be set to not be readable)
+    assumeTrue(dummy.setReadable(false));
 
     // then
     // System.out.println("was set = " + isSet);

--- a/broker-core/src/test/java/io/zeebe/broker/system/ConfigurationTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/system/ConfigurationTest.java
@@ -31,12 +31,17 @@ import io.zeebe.broker.system.configuration.SocketBindingSubscriptionCfg;
 import io.zeebe.broker.system.configuration.TomlConfigurationReader;
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
+import java.nio.file.Paths;
 import java.util.HashMap;
 import java.util.Map;
 import org.assertj.core.api.Condition;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
 
 public class ConfigurationTest {
+
+  @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
 
   public Map<String, String> environment = new HashMap<>();
 
@@ -124,7 +129,8 @@ public class ConfigurationTest {
                     + "id=\"internal-2\"")
                 .getBytes());
     final BrokerCfg config = TomlConfigurationReader.read(input);
-    final String base = "/opt/zeebe";
+    final String base = temporaryFolder.getRoot().getAbsolutePath();
+    final String jarFile = Paths.get(base, "exporters", "exporter.jar").toAbsolutePath().toString();
 
     // when
     config.init(base);
@@ -132,7 +138,7 @@ public class ConfigurationTest {
     // then
     assertThat(config.getExporters()).hasSize(3);
     assertThat(config.getExporters().get(0))
-        .hasFieldOrPropertyWithValue("jarPath", base + "/exporters/exporter.jar")
+        .hasFieldOrPropertyWithValue("jarPath", jarFile)
         .is(new Condition<>(ExporterCfg::isExternal, "is external"));
     assertThat(config.getExporters().get(1).isExternal()).isFalse();
     assertThat(config.getExporters().get(2).isExternal()).isFalse();

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogSegment.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogSegment.java
@@ -104,9 +104,7 @@ public class FsLogSegment {
 
   public void delete() {
     final File file = new File(fileName);
-    if (file.exists()) {
-      file.delete();
-    }
+    FileUtil.deleteFile(file);
   }
 
   public String getFileName() {

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogStorage.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogStorage.java
@@ -362,10 +362,7 @@ public class FsLogStorage implements LogStorage {
   protected void deleteBackupFilesIfExist(File logDir) {
     final List<File> backupFiles =
         Arrays.asList(logDir.listFiles(config::matchesBackupFileNamePattern));
-    backupFiles.forEach(
-        (file) -> {
-          file.delete();
-        });
+    backupFiles.forEach(FileUtil::deleteFile);
   }
 
   protected void applyTruncatedFileIfExists(File logDir) {
@@ -380,10 +377,7 @@ public class FsLogStorage implements LogStorage {
       if (shouldApplyTruncatedSegment(logDir, truncatedFile, truncatedSegmentId)) {
         moveFile(truncatedFile.getAbsolutePath(), config.fileName(truncatedSegmentId));
       } else {
-        truncatedFiles.forEach(
-            (file) -> {
-              file.delete();
-            });
+        truncatedFiles.forEach(FileUtil::deleteFile);
       }
 
     } else if (truncatedApplicableFiles > 1) {

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsReadableSnapshot.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsReadableSnapshot.java
@@ -80,8 +80,8 @@ public class FsReadableSnapshot extends FsSnapshotMetadata implements ReadableSn
   public void delete() {
     FileUtil.closeSilently(inputStream);
 
-    dataFile.delete();
-    checksumFile.delete();
+    FileUtil.deleteFile(dataFile);
+    FileUtil.deleteFile(checksumFile);
   }
 
   @Override

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsSnapshotStorage.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsSnapshotStorage.java
@@ -18,6 +18,7 @@ package io.zeebe.logstreams.impl.snapshot.fs;
 import io.zeebe.logstreams.impl.Loggers;
 import io.zeebe.logstreams.spi.SnapshotMetadata;
 import io.zeebe.logstreams.spi.SnapshotStorage;
+import io.zeebe.util.FileUtil;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
@@ -74,8 +75,8 @@ public class FsSnapshotStorage implements SnapshotStorage {
     boolean deletionSuccessful = false;
     if (snapshotFiles.size() > 0) {
       for (File snapshotFile : snapshotFiles) {
-        getChecksumFile(snapshotFile, name).delete();
-        snapshotFile.delete();
+        FileUtil.deleteFile(getChecksumFile(snapshotFile, name));
+        FileUtil.deleteFile(snapshotFile);
       }
 
       deletionSuccessful = true;
@@ -102,7 +103,7 @@ public class FsSnapshotStorage implements SnapshotStorage {
     try {
       snapshotFile.createNewFile();
     } catch (IOException e) {
-      snapshotFile.delete();
+      FileUtil.deleteFile(snapshotFile);
       throw e;
     }
 

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsSnapshotWriter.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsSnapshotWriter.java
@@ -94,8 +94,8 @@ public class FsSnapshotWriter implements SnapshotWriter {
   @Override
   public void abort() {
     FileUtil.closeSilently(dataOutputStream);
-    dataFile.delete();
-    checksumFile.delete();
+    FileUtil.deleteFile(dataFile);
+    FileUtil.deleteFile(checksumFile);
   }
 
   public File getChecksumFile() {
@@ -141,7 +141,7 @@ public class FsSnapshotWriter implements SnapshotWriter {
 
     try (FileOutputStream checksumFileOutputStream = new FileOutputStream(checksumFile);
         BufferedOutputStream checksumOutputStream =
-            new BufferedOutputStream(checksumFileOutputStream); ) {
+            new BufferedOutputStream(checksumFileOutputStream)) {
       final String checksumHex = BitUtil.toHex(checksum);
       final String checksumFileContents = config.checksumContent(checksumHex, getDataFileName());
 

--- a/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsTemporarySnapshotWriter.java
+++ b/logstreams/src/main/java/io/zeebe/logstreams/impl/snapshot/fs/FsTemporarySnapshotWriter.java
@@ -43,7 +43,7 @@ public class FsTemporarySnapshotWriter extends FsSnapshotWriter {
     try {
       writeChecksumFile(checksum);
       FileUtil.replace(dataFile.toPath(), snapshotFile.toPath());
-      dataFile.delete();
+      FileUtil.deleteFile(dataFile);
     } catch (final Exception ex) {
       abort();
       throw ex;
@@ -55,7 +55,7 @@ public class FsTemporarySnapshotWriter extends FsSnapshotWriter {
   @Override
   public void abort() {
     super.abort();
-    snapshotFile.delete();
+    FileUtil.deleteFile(snapshotFile);
   }
 
   @Override

--- a/logstreams/src/test/java/io/zeebe/logstreams/fs/snapshot/FsSnapshotStorageTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/fs/snapshot/FsSnapshotStorageTest.java
@@ -25,6 +25,7 @@ import io.zeebe.logstreams.impl.snapshot.fs.FsSnapshotStorageConfiguration;
 import io.zeebe.logstreams.impl.snapshot.fs.FsSnapshotWriter;
 import io.zeebe.logstreams.impl.snapshot.fs.FsTemporarySnapshotWriter;
 import io.zeebe.logstreams.spi.SnapshotMetadata;
+import io.zeebe.util.FileUtil;
 import java.io.File;
 import java.io.IOException;
 import java.io.OutputStream;
@@ -142,7 +143,7 @@ public class FsSnapshotStorageTest {
     fsSnapshotWriter.getOutputStream().write(SNAPSHOT_DATA);
     fsSnapshotWriter.commit();
 
-    fsSnapshotWriter.getChecksumFile().delete();
+    FileUtil.deleteFile(fsSnapshotWriter.getChecksumFile());
 
     final FsReadableSnapshot snapshot = fsSnapshotStorage.getLastSnapshot("test");
     assertThat(snapshot).isNull();

--- a/logstreams/src/test/java/io/zeebe/logstreams/fs/snapshot/FsTemporarySnapshotWriterTest.java
+++ b/logstreams/src/test/java/io/zeebe/logstreams/fs/snapshot/FsTemporarySnapshotWriterTest.java
@@ -19,7 +19,6 @@ import static io.zeebe.util.StringUtil.getBytes;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
-import io.zeebe.UnstableTest;
 import io.zeebe.logstreams.impl.snapshot.fs.FsReadableSnapshot;
 import io.zeebe.logstreams.impl.snapshot.fs.FsSnapshotStorage;
 import io.zeebe.logstreams.impl.snapshot.fs.FsSnapshotStorageConfiguration;
@@ -32,7 +31,6 @@ import org.agrona.BitUtil;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
-import org.junit.experimental.categories.Category;
 import org.junit.rules.TemporaryFolder;
 
 public class FsTemporarySnapshotWriterTest {
@@ -114,13 +112,9 @@ public class FsTemporarySnapshotWriterTest {
   }
 
   @Test
-  @Category(UnstableTest.class)
   public void shouldStillHaveLastSnapshotIfFailToMoveTemporaryFile() throws Exception {
     // given
     final byte[] badChecksum = new byte[] {0};
-    final FsTemporarySnapshotWriter writer =
-        new FsTemporarySnapshotWriter(
-            config, temporaryFile, checksumFile, snapshotFile, lastSnapshot);
 
     // when
     writer.getOutputStream().write(SNAPSHOT_DATA);

--- a/logstreams/src/test/jmh/io/zeebe/logstreams/snapshot/benchmarks/FilledMapSnapshotSupplier.java
+++ b/logstreams/src/test/jmh/io/zeebe/logstreams/snapshot/benchmarks/FilledMapSnapshotSupplier.java
@@ -20,6 +20,7 @@ import static io.zeebe.map.ZbMap.DEFAULT_BLOCK_COUNT;
 import io.zeebe.logstreams.snapshot.ComposedSnapshot;
 import io.zeebe.logstreams.snapshot.ZbMapSnapshotSupport;
 import io.zeebe.map.Long2LongZbMap;
+import io.zeebe.util.FileUtil;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Files;
@@ -53,6 +54,6 @@ public class FilledMapSnapshotSupplier {
   @TearDown(Level.Iteration)
   public void closeMap() {
     map.close();
-    tmpFile.delete();
+    FileUtil.deleteFile(tmpFile);
   }
 }

--- a/logstreams/src/test/jmh/io/zeebe/logstreams/snapshot/benchmarks/WrittenMapSnapshotSupplier.java
+++ b/logstreams/src/test/jmh/io/zeebe/logstreams/snapshot/benchmarks/WrittenMapSnapshotSupplier.java
@@ -20,6 +20,7 @@ import static io.zeebe.map.ZbMap.DEFAULT_BLOCK_COUNT;
 import io.zeebe.logstreams.snapshot.ComposedSnapshot;
 import io.zeebe.logstreams.snapshot.ZbMapSnapshotSupport;
 import io.zeebe.map.Long2LongZbMap;
+import io.zeebe.util.FileUtil;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.util.Random;
@@ -54,6 +55,6 @@ public class WrittenMapSnapshotSupplier {
   @TearDown(Level.Iteration)
   public void closeMap() {
     map.close();
-    tmpFile.delete();
+    FileUtil.deleteFile(tmpFile);
   }
 }

--- a/util/src/main/java/io/zeebe/util/FileUtil.java
+++ b/util/src/main/java/io/zeebe/util/FileUtil.java
@@ -33,8 +33,11 @@ import java.nio.file.Paths;
 import java.nio.file.SimpleFileVisitor;
 import java.nio.file.attribute.BasicFileAttributes;
 import org.agrona.LangUtil;
+import org.slf4j.Logger;
 
 public class FileUtil {
+
+  public static final Logger LOG = Loggers.FILE_LOGGER;
 
   public static void closeSilently(Closeable out) {
     if (out != null) {
@@ -159,5 +162,11 @@ public class FileUtil {
     }
 
     return path;
+  }
+
+  public static void deleteFile(File file) {
+    if (file.exists() && !file.delete()) {
+      LOG.warn("Failed to delete file '{}'", file);
+    }
   }
 }

--- a/util/src/main/java/io/zeebe/util/Loggers.java
+++ b/util/src/main/java/io/zeebe/util/Loggers.java
@@ -21,4 +21,5 @@ public class Loggers {
 
   public static final Logger ACTOR_LOGGER = new ZbLogger("io.zeebe.util.actor");
   public static final Logger IO_LOGGER = new ZbLogger("io.zeebe.util.buffer");
+  public static final Logger FILE_LOGGER = new ZbLogger("io.zeebe.util.fs");
 }

--- a/zb-map/src/test/java/io/zeebe/map/BucketBufferArrayIOTest.java
+++ b/zb-map/src/test/java/io/zeebe/map/BucketBufferArrayIOTest.java
@@ -40,6 +40,7 @@ import io.zeebe.map.types.LongValueHandler;
 import io.zeebe.test.util.io.AlwaysFailingInputStream;
 import io.zeebe.test.util.io.RepeatedlyFailingInputStream;
 import io.zeebe.test.util.io.RepeatedlyFailingOutputStream;
+import io.zeebe.util.FileUtil;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
@@ -297,7 +298,7 @@ public class BucketBufferArrayIOTest {
     final File tmpFile = File.createTempFile("tmpFile", ".tmp");
     final FileOutputStream outputStream = new FileOutputStream(tmpFile);
     outputStream.close();
-    tmpFile.delete();
+    FileUtil.deleteFile(tmpFile);
 
     // expect IOException
     expectedException.expect(IOException.class);

